File: scripts/aux_scripts/get_negatives_from_disease.rb
#! /usr/bin/env ruby
require 'optparse'

########################### FUNCTIONS #######################
#############################################################

def write_negatives(file, negatives)
  File.open(file ,'w') do |f|
    negatives.each do |group, nodes|
      f.puts "#{group}" + "\t" + nodes.join(",")
    end
  end
end

def get_negatives2groups(disgroup_genes)
  all_genes = disgroup_genes.values.flatten.uniq
  negatives = {}
  disgroup_genes.each_key do |k|
    negatives[k] =  all_genes - disgroup_genes[k].uniq
  end
  return negatives
end

def get_negatives2disease(diseases_disgroup,disgroup_negatives)
  negatives = {}
  diseases_disgroup.each do |disease, disgroup|
    negatives[disease] = disgroup_negatives[disgroup]
  end
  return negatives
end


def load_node_groups_from_file(file, sep: ',')
   diseases = {}
   disgroup_genes = {}
   File.open(file).each do |line|
     disease_name, disgroup, genes = line.chomp.split("\t")
     diseases[disease_name] = disgroup
     if disgroup_genes[disgroup].nil?
       disgroup_genes[disgroup] = genes.split(sep)
     else
       disgroup_genes[disgroup] += genes.split(sep)
     end
   end

   return diseases, disgroup_genes
 end


########################### OPTPARSE ########################
#############################################################

options = {}
OptionParser.new do  |opts|

  options[:positive_file] = nil
  opts.on("-i","-input_positives POS", "The roots for the positive file") do |item|
    options[:positive_file] = item
  end

  options[:output_name] = "negatives"
  opts.on("-o","-output_name NAME", "The name of the ranked file") do |output_name|
    options[:output_name] = output_name
  end

end.parse!

########################### MAIN ############################
#############################################################

diseases_disgroup, disgroup_genes = load_node_groups_from_file(options[:positive_file])
disgroup_negatives = get_negatives2groups(disgroup_genes)
diseases_negatives = get_negatives2disease(diseases_disgroup,disgroup_negatives)
write_negatives(options[:output_name],diseases_negatives) if !diseases_negatives.nil?

File: scripts/aux_scripts/get_negatives_from_disease.py
#! /usr/bin/env python
import argparse
import numpy as np 

########################### FUNCTIONS #######################
#############################################################

def write_negatives(file, negatives):
	with open(file, "w") as f:
		for group, nodes in negatives.items():
			f.write(group + "\t" + ",".join(nodes) + "\n")


def get_negatives2groups(disgroup_genes):
	all_genes = set([el for row in disgroup_genes.values() for el in row])
	negatives = {}
	for k, v in disgroup_genes.items():
		number_of_positives = len(set(v))
		number_of_negatives = int(np.floor(number_of_positives / 2))
		smpl_negatives = list(all_genes - set(v))
		smpl_negatives = list(np.random.choice(smpl_negatives, number_of_negatives, replace=False))
		negatives[k] = smpl_negatives
	return negatives


def load_node_groups_from_file(file, sep= ','):
	disgroup_genes = {}
	with open(file) as f:
		for line in f:
			disgroup, genes = line.strip().split("\t")
			disgroup_genes[disgroup] = genes.split(sep)
	return disgroup_genes


########################### OPTPARSE ########################
#############################################################

parser = argparse.ArgumentParser(description='Process some integers.')

parser.add_argument('-i', '--input_positives', required=True,
					help='The roots for the positive file')

parser.add_argument('-o', '--output_name', default='negatives',
					help='The name of the ranked file')

parser.add_argument("--subset_randomly_negatives", default=True, action= "store_false", help=" Select this to choos all background as negatives")

options = parser.parse_args()

########################### MAIN ############################
#############################################################

positive_file = options.input_positives
output_name = options.output_name

disgroup_genes = load_node_groups_from_file(positive_file)
disgroup_negatives = get_negatives2groups(disgroup_genes)

if disgroup_negatives  is not None:
	write_negatives(output_name, disgroup_negatives )